langchain/retrievers/document_compressors/chain_extract.py

```diff
@@ -49,19 +49,29 @@ def compress_documents(
         self, documents: Sequence[Document], query: str
     ) -> Sequence[Document]:
         """Compress page content of raw documents."""
-        compressed_docs = []
-        for doc in documents:
-            _input = self.get_input(query, doc)
-            output = self.llm_chain.predict_and_parse(**_input)
-            if len(output) == 0:
-                continue
-            compressed_docs.append(Document(page_content=output, metadata=doc.metadata))
+        outputs = self.llm_chain.apply_and_parse(
+            [self.get_input(query, doc) for doc in documents]
+        )
+        compressed_docs = [
+            Document(page_content=output, metadata=doc.metadata)
+            for output, doc in zip(outputs, documents)
+            if len(output) > 0
+        ]
         return compressed_docs
 
     async def acompress_documents(
         self, documents: Sequence[Document], query: str
     ) -> Sequence[Document]:
-        raise NotImplementedError
+        """Compress page content of raw documents."""
+        outputs = await self.llm_chain.aapply_and_parse(
```

Comment:
@vowelparrot As mentioned in the PR, I basically copied the implementation from the `MapReduceDocumentsChain`. As far as I understand, the `apply`/`aapply` methods are specifically created to apply chain operations to sequences or multiple inputs/documents, which is exactly the case here. 

Thus, imho there are two advantages of this change:

- The implementations for `compress_documents` and `acompress_documents` are basically identical which increases maintainability for changes to sync/async operations in the future.
- due to the similar syntax, it will be easier to wrap some existing CombineDocumentsChain for LLMChainExtractor instead of maintaining a separate DocumentCompressor implementation (but that is an architetctural decision).

But if you prefer that, I can revert the sync changes and just add the async implementation; this is my first contribution so  happy to get guidance.....

Reply:
yeah lets revert the sync changes and just add async!